src/Umbraco.Core/Mapping/Mapper.cs

```diff
@@ -0,0 +1,366 @@
+        /// <param name="source">The source object.</param>
+        /// <param name="context">A mapper context.</param>
+        /// <returns>The target object.</returns>
+        public TTarget Map<TSource, TTarget>(TSource source, MapperContext context)
+            => Map<TTarget>(source, typeof(TSource), context);
+
+        private TTarget Map<TTarget>(object source, Type sourceType, MapperContext context)
+        {
+            if (source == null)
+                throw new ArgumentNullException(nameof(source));
+
+            var targetType = typeof(TTarget);
+
+            var ctor = GetCtor(sourceType, targetType);
+            var map = GetMap(sourceType, targetType);
+
+            // if there is a direct constructor, map
+            if (ctor != null && map != null)
+            {
+                var target = ctor(source, context);
+                map(source, target, context);
+                return (TTarget)target;
+            }
+
+            // otherwise, see if we can deal with enumerable
+
+            var ienumerableOfT = typeof(IEnumerable<>);
+
+            bool IsIEnumerableOfT(Type type) =>
+                type.IsGenericType &&
+                type.GenericTypeArguments.Length == 1 &&
+                type.GetGenericTypeDefinition() == ienumerableOfT;
+
+            // try to get source as an IEnumerable<T>
+            var sourceIEnumerable = IsIEnumerableOfT(sourceType) ? sourceType : sourceType.GetInterfaces().FirstOrDefault(IsIEnumerableOfT);
+
+            // if source is an IEnumerable<T> and target is T[] or IEnumerable<T>, we can create a map
+            if (sourceIEnumerable != null && IsEnumerableOrArrayOfType(targetType))
+            {
+                var sourceGenericArg = sourceIEnumerable.GenericTypeArguments[0];
+                var targetGenericArg = GetEnumerableOrArrayTypeArgument(targetType);
+
+                ctor = GetCtor(sourceGenericArg, targetGenericArg);
+                map = GetMap(sourceGenericArg, targetGenericArg);
+
+                // if there is a constructor for the underlying type, create & invoke the map
+                if (ctor != null && map != null)
+                {
+                    // register (for next time) and do it now (for this time)
+                    object NCtor(object s, MapperContext c) => MapEnumerable<TTarget>((IEnumerable) s, targetGenericArg, ctor, map, c);
+                    DefineCtors(sourceType)[targetType] = NCtor;
+                    DefineMaps(sourceType)[targetType] = Identity;
+                    return (TTarget) NCtor(source, context);
+                }
+
+                throw new InvalidOperationException($"Don't know how to map {sourceGenericArg.FullName} to {targetGenericArg.FullName}, so don't know how to map {sourceType.FullName} to {targetType.FullName}.");
+            }
+
+            throw new InvalidOperationException($"Don't know how to map {sourceType.FullName} to {targetType.FullName}.");
```

Comment:
There are no multiple reason. The only reason is: we don't know how to map, because no map has been defined. It's not that we tried and failed - it's that we were never told how to do it in the first place. So, no, cannot be more detailed.